doc/dev/python_scheduling/collect_pointwise_data.py

```diff
@@ -0,0 +1,253 @@
+def run_profile(presched_fd, inputs, config=None):
+    scheduled_fd = custom_pointwise_scheduler(presched_fd, config)
+    outputs = scheduled_fd.execute(inputs, profile=True)
+
+    # validate correctness
+    # TODO use thunder to compile torch eager to nvfuser fusion
+    # assert torch.allclose(nvf_outputs[0], eager_reference(inputs))
+
+    prof = scheduled_fd.profile()
+
+    num_registers = prof.kernel_profiles[0].registers
+    smem = prof.kernel_profiles[0].shared_mem_str
+    grid_shape = prof.kernel_profiles[0].grid_str
+    block_shape = prof.kernel_profiles[0].block_str
+    bandwidth = prof.kernel_profiles[0].effective_bandwidth_gbs
+    time = prof.kernel_profiles[0].time_ms
+
+    return outputs, (grid_shape, block_shape, num_registers, smem, bandwidth, time)
+
+
+# ============================ Create Fusion  ================================
+
+
+def generate_shapes_with_bcast(tensor_shape):
+    num_dims = len(tensor_shape)
+    for num_bcast_dims in range(num_dims):
+        # Get all combinations of broadcast dimensions
+        for bcast_indices in itertools.combinations(range(num_dims), num_bcast_dims):
+            yield [
+                dim_size if idx not in bcast_indices else 1
+                for idx, dim_size in enumerate(tensor_shape)
+            ]
+
+
+def create_fusion_state(full_tensor_shape, maximum_number_operations):
+    # vertical composition: number of operations, order of mufu operations
+    # horizontal composition: number and size of tensor dimensions, order of
+    # broadcast dimensions.
+
+    for num_ops in range(1, maximum_number_operations + 1):
+        for num_mufu in range(num_ops + 1):
+            # Get a random subset of num_mufu indices from num_ops
+            mufu_indices = np.random.choice(num_ops, num_mufu, replace=False)
+
+            num_tensors = 1 + num_ops - num_mufu
+            tensor_shapes_with_bcast = list(
+                generate_shapes_with_bcast(full_tensor_shape)
+            )
+            shapes_for_all_tensors = [tensor_shapes_with_bcast] * num_tensors
+            for input_shapes in itertools.product(*shapes_for_all_tensors):
+                yield (num_ops, mufu_indices, input_shapes)
+
+
+def create_fusion_definition(num_operations, mufu_indices, input_shapes):
+    input_tensors = [torch.randn(shape, device="cuda") for shape in input_shapes]
+
+    with FusionDefinition() as fd:
+        output = None
+        output_tensor = None
+        input_tensor_idx = 0
+        for idx in range(num_operations):
+            if output is None:
+                output = fd.from_pytorch(input_tensors[input_tensor_idx])
+                input_tensor_idx += 1
+
+            if idx in mufu_indices:
+                output = fd.ops.exp(output)
+            else:
+                other = fd.from_pytorch(input_tensors[input_tensor_idx])
+                input_tensor_idx += 1
+                output = fd.ops.add(output, other)
+        fd.add_output(output)
```

Comment:
should we explore more topology rather than always chained operations?

Reply:
Good idea. At the moment, I was working backwards from the information available to the Pointwise scheduler. It doesn't compute any metrics about the fusion's DAG.

Reply:
Is a more complex topology relevant for pointwise scheduler? e.g., The DAG is flattened to a linear sequence of operations, so it does not affect the arithmetic intensity of the Fusion.